src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java

```diff
@@ -0,0 +1,212 @@
+ * the false sharing problem.
+ */
+public class HybridJacksonPool implements RecyclerPool<BufferRecycler> {
+
+  private static final RecyclerPool INSTANCE = new HybridJacksonPool();
+
+  private static final Predicate<Thread> isVirtual = VirtualPredicate.findIsVirtualPredicate();
+
+  private final RecyclerPool<BufferRecycler> nativePool = JsonRecyclerPools.threadLocalPool();
+
+  static class VirtualPoolHolder {
+    // Lazy on-demand initialization
+    private static final RecyclerPool<BufferRecycler> virtualPool = new StripedLockFreePool(Runtime.getRuntime().availableProcessors());
+  }
+
+  public static RecyclerPool getInstance() {
+    return INSTANCE;
+  }
+
+  @Override
+  public BufferRecycler acquirePooled() {
+    return isVirtual.test(Thread.currentThread()) ? VirtualPoolHolder.virtualPool.acquirePooled()
+      : nativePool.acquirePooled();
+  }
+
+  @Override
+  public void releasePooled(BufferRecycler bufferRecycler) {
+    if (bufferRecycler instanceof VThreadBufferRecycler) {
+      // if it is a PooledBufferRecycler it has been acquired by a virtual thread, so it has to be release to the same pool
+      VirtualPoolHolder.virtualPool.releasePooled(bufferRecycler);
+    }
+    // the native thread pool is based on ThreadLocal, so it doesn't have anything to do on release
+  }
+
+  private static class StripedLockFreePool implements RecyclerPool<BufferRecycler> {
+
+    private static final int CACHE_LINE_SHIFT = 4;
+
+    private static final int CACHE_LINE_PADDING = 1 << CACHE_LINE_SHIFT;
+
+    private final XorShiftThreadProbe threadProbe;
+
+    private final AtomicReferenceArray<Node> heads;
+
+    public StripedLockFreePool(int stripesCount) {
+      if (stripesCount <= 0) {
+        throw new IllegalArgumentException("Expecting a stripesCount that is larger than 0");
+      }
+
+      int size = roundToPowerOfTwo(stripesCount);
+      this.heads = new AtomicReferenceArray<>(size * CACHE_LINE_PADDING);
+
+      int mask = (size - 1) << CACHE_LINE_SHIFT;
+      this.threadProbe = new XorShiftThreadProbe(mask);
+    }
+
+    @Override
+    public BufferRecycler acquirePooled() {
+      int index = threadProbe.index();
+
+      Node currentHead = heads.get(index);
+      while (true) {
+        if (currentHead == null) {
+          return new VThreadBufferRecycler(index);
+        }
+
+        if (heads.compareAndSet(index, currentHead, currentHead.next)) {
+          currentHead.next = null;
+          return currentHead.value;
+        } else {
+          currentHead = heads.get(index);
+        }
+      }
+    }
+
+    @Override
+    public void releasePooled(BufferRecycler recycler) {
+      VThreadBufferRecycler vThreadBufferRecycler = (VThreadBufferRecycler) recycler;
+      Node newHead = new Node(vThreadBufferRecycler);
+
+      Node next = heads.get(vThreadBufferRecycler.slot);
+      while (true) {
+        if (heads.compareAndSet(vThreadBufferRecycler.slot, next, newHead)) {
```

Comment:
right before this: `newHead.level = next.level + 1` - level doesn't need to be volatile, bacause is correctly acquired/released by reading the `Node`'s instance and performing this `compareAndset`.